src/infra/core/src/compact_service_impl.rs

```diff
@@ -0,0 +1,386 @@
+
+        let chain = dataset.as_metadata_chain();
+        let head = chain.resolve_ref(&BlockRef::Head).await?;
+        let mut block_stream = chain.iter_blocks_interval(&head, None, false);
+        let object_data_repo = dataset.as_data_repo();
+        let object_block_repo = chain.as_metadata_block_repository();
+
+        while let Some((block_hash, block)) = block_stream.try_next().await? {
+            block_file_urls.push(object_block_repo.get_internal_url(&block_hash).await);
+
+            match block.event {
+                MetadataEvent::AddData(add_data_event) => {
+                    if let Some(output_slice) = &add_data_event.new_data {
+                        new_head = block.prev_block_hash.clone();
+
+                        let data_slice_url = object_data_repo
+                            .get_internal_url(&output_slice.physical_hash)
+                            .await;
+                        if data_slice_batch_info.data_slices_batch.is_empty() {
+                            new_end_offset_interval = output_slice.offset_interval.end;
+                        }
+
+                        let current_records = output_slice.offset_interval.end
+                            - output_slice.offset_interval.start
+                            + 1;
+
+                        if batch_size + output_slice.size >= max_slice_size
+                            || batch_records + current_records >= max_slice_records
+                        {
+                            if !data_slice_batch_info.data_slices_batch.is_empty() {
+                                data_slice_batches.push(data_slice_batch_info.clone());
+                                // Reset values for next batch
+                                data_slice_batch_info.new_file_path = None;
+                                data_slice_batch_info.new_offset_interval = None;
+                                new_end_offset_interval = output_slice.offset_interval.end;
+                            }
+
+                            data_slice_batch_info.data_slices_batch = vec![data_slice_url];
+                            batch_size = output_slice.size;
+                            batch_records = current_records;
+                        } else {
+                            data_slice_batch_info.data_slices_batch.push(data_slice_url);
+                            batch_size += output_slice.size;
+                            batch_records += current_records;
+                        }
+
+                        if data_slice_batch_info.new_file_path.is_none() {
+                            data_slice_batch_info.new_file_path = Some(compact_dir_path.join(
+                                Self::get_random_operation_name_with_prefix(
+                                    format!("{dataset_name}-").as_str(),
+                                ),
+                            ));
+                        }
```

Comment:
This looks unrelated to the gathering algorithm. This is already an operation assignment, post-processing de-facto. Maybe it should be done separately or event does not have to be stored. I think it should be generated in `write_file` and returned to use in `replace_data_files` instead of being part of this struct.